muikku-core-plugins/src/main/java/fi/otavanopisto/muikku/plugins/workspacenotes/rest/WorkspaceNoteRESTService.java

```diff
@@ -0,0 +1,323 @@
+   *  
+   *  Parameter rest model must contain owner, workspaceEntityId & nextSiblingId. 
+   *  
+   *  example:
+   *  {owner: 14, workspaceEntityId: 23, title: "Title", workspaceNote: "updatedNote", nextSiblingId: 2}
+   *  
+   *  
+   *  Editable fields are only title, workspaceNote & nextSiblingId
+   *  
+   *  returns updated rest model
+   *  
+   *  WorkspaceNoteRestModel: = { 
+   *    id: 1,
+   *    title: "updated title here",
+   *    workspaceNote: "updated text here",
+   *    workspaceEntityId: 123,
+   *    owner: 1
+   * };
+   *  
+   *  Errors:
+   *  404 Not found if workspaceNote not found
+   *  400 Bad request if userEntityId is null
+   *  403 Forbidden if userEntityId does not match with logged user
+   */
+  @PUT
+  @Path ("/workspacenote")
+  @RESTPermit (handling = Handling.INLINE, requireLoggedIn = true)
+  public Response updateWorkspaceNote(WorkspaceNoteRestModel restModel) {
+    
+    WorkspaceNote workspaceNote = workspaceNoteController.findWorkspaceNoteById(restModel.getId());
+    
+    if (workspaceNote == null) {
+      return Response.status(Status.NOT_FOUND).build();
+    }
+    
+    if (restModel.getOwner() == null) {
+      return Response.status(Status.BAD_REQUEST).build();
+    } 
+    
+    if(restModel.getWorkspaceEntityId() == null || !restModel.getWorkspaceEntityId().equals(workspaceNote.getWorkspace())) {
+      return Response.status(Status.BAD_REQUEST).entity("Workspace entity id mismatch").build();
+    }
+    
+    // users can only update their own notes (except admins)    
+    if (!restModel.getOwner().equals(sessionController.getLoggedUserEntity().getId())) {
+      EnvironmentRoleArchetype loggedUserRole = getUserRoleArchetype(sessionController.getLoggedUser());
+
+      if (!loggedUserRole.equals(EnvironmentRoleArchetype.ADMINISTRATOR)) {
+        return Response.status(Status.FORBIDDEN).build();
+      }
+    }
+    WorkspaceNote referenceNote = null;
+    
+    if (restModel.getNextSiblingId() != null) {
+      referenceNote = workspaceNoteController.findWorkspaceNoteById(restModel.getNextSiblingId());
+    }
+    
+    if (referenceNote == null) {
+      Integer maximumOrderNumber = workspaceNoteController.getMaximumOrderNumberByOwnerAndWorkspace(restModel.getWorkspaceEntityId(), restModel.getOwner());
+      workspaceNote = workspaceNoteController.updateOrderNumber(workspaceNote, ++maximumOrderNumber);
+    } else {
+      workspaceNote = workspaceNoteController.moveAbove(workspaceNote, referenceNote);
+    }
+    
+    workspaceNote = workspaceNoteController.updateWorkspaceNote(workspaceNote, restModel.getTitle(), restModel.getWorkspaceNote());
+    
+    return Response.ok(toRestModel(workspaceNote)).build();
+  }
+  
+  private WorkspaceNoteRestModel toRestModel(WorkspaceNote workspaceNote) {
+    
+    UserEntity userEntity = userEntityController.findUserEntityById(workspaceNote.getOwner());
```

Comment:
Is userEntity needed here for anything else than userEntity.getId (line 184) which is equal to workspaceNote.getOwner? Could skip this step.